Feature/enhanced columns

binarystargames opened the pull request:
Added a "ratio" feature to columns shortcode. Code sample/execution:

```
{{< flexcolumns ratio="3:7" >}}
**Sunrod**<br/>
*Bomb, At-Will, Standard*<br/>
*Spend 1 Reagent*
<--->
<br/>
Make an attack vs PD on 1d3+1 nearby targets.

*Hit:* The target takes radiant damage equal to your level and is Dazed (11+ save ends).

*Miss:* The target is Dazed until the start of your next turn.

*Special:* You may partially activate this as a free action (including outside of combat) to create light twice as bright as a torch for about an hour. While still lit, you may throw it to use the Bomb ability as above (it flares up and then stops working on contact).

{{< /flexcolumns >}}
```

This code becomes: https://binarystar.games/docs/36th-way/classes/list-of-classes/alchemist/#1st-level

---

I've also got a CSS rule locally to flatten columns to vertical when hugo-book gets rid of the side TOC/menu because the default hugo-book ones look not great at low width. I left it out for now but it's an option.

binarystargames committed:
Add flexcolumns shortcode

Prior to this commit, theme users were able to specify the `columns`
shortcode to arrange their content into multiple evenly distributed
columns.

This commit builds on that work by introducing a new `flexcolumns`
shortcode which allows users to optionally specify a ratio to have more
control over column widths.

michaeltlombardi commented:
Looking over this, I wonder if this isn't worth contributing back upstream to `hugo-book` directly as it's explicitly an enhancement of their shortcode. If we file an issue with them to enhance the `columns` shortcode with the `ratio` parameter, I think that will be strictly better for everyone.

If they reject the feature request / implementation, I think we can instead incorporate this as `{{ flex-columns }}` or something; I think the changeset for this is (without refactoring) too large to feasibly keep patched if the upstream implementation changes. That being said, this file does not seem to update often; 3 patches since implementation, only one since 2020.

binarystargames commented:
Makes sense. I'll see if I have the energy to look into their requested process sometime this week.

michaeltlombardi commented:
If you want, I'm happy to go ahead and file the issue and link to your prototype here - I really do think this is a great improvement over the upstream shortcode and hugely useful for folks!

binarystargames commented:
That would be very appreciated!

michaeltlombardi commented:
@binarystargames I think the upstream maintainer is unlikely to merge your contribution to the upstream. Instead of having to maintain our own implementation on top of theirs, can this be reworked to be a new shortcode called `flexcolumns`? And then we can document it and just have it be an extension here instead.

binarystargames commented:
Yep, I can do that.

binarystargames commented:
Added a commit that does just that, updated the overview example. I implemented it locally on my site, so that example link is current - that's using the new `{{< flexcolumns ratio="3:7" >}}` situation.

binarystargames commented:
Secondary question: while I'm working on columns, how do you feel about adding a parameter to flatten them whenever the sidebar disappears? We can certainly leave it to CSS to handle too, which is what I'm currently doing for my site. Not sure how you feel about one implementation vs. the other.

michaeltlombardi commented on layouts/shortcodes/flexcolumns.html:
```diff
@@ -0,0 +1,18 @@
+  </div>
+  {{ end }}
+{{ end }}
+</div>
```
nit: please add a trailing newline

binarystargames replied:
Done (and I added it to the other shortcodes).

michaeltlombardi commented:
> Secondary question: while I'm working on columns, how do you feel about adding a parameter to flatten them whenever the sidebar disappears? We can certainly leave it to CSS to handle too, which is what I'm currently doing for my site. Not sure how you feel about one implementation vs. the other.

I think having a shortcode parameter like "flattenInMobileView" would be very useful - if people want more control over the presentation, they can ignore that parameter flag and implement their own CSS, but I don't think that's the _majority_ of our target audience.

binarystargames commented:
Done. 

To accomplish this, I also added `platen.scss` injected via `inject/head.html` (because I suspect to do at least a few of the things on that issues tracker we'll want somewhere to put this kind of stuff). Let me know if you'd rather I separate that out to another PR or something.

By default it uses default columns behavior, but including `flattenInMobileView="true"` adds the flattening. (I added it to Sunrod and Tanglefoot Bag from the initial demo link and nothing else so you can see the difference: https://binarystar.games/docs/36th-way/classes/list-of-classes/alchemist/#1st-level )

michaeltlombardi commented on layouts/shortcodes/flexcolumns.html:
```diff
@@ -0,0 +1,22 @@
+{{ $mobile := "" }}
```
Something I find it useful to do for shortcodes is get/declare the variables at the top of the file to help clarify what the shortcode can expect to receive from a user / allow us to just reference the variables further down (unless they have to be pulled from the current scope, like `$ratioSplit` does). That also lets us handle when the shortcodes are called without named parameters.

You can such an example here:

https://github.com/platenio/hugo-platen/blob/64d2e468e33b53a3c14130bfe3561ba92bea5f36/layouts/shortcodes/itchio.html#L1-L11

We can _also_ use the [`default` function](https://gohugo.io/functions/default/) to set any defaults we need during parameter declaration. We could possibly use this mechanism to define a default ratio to keep from forking the code later (or not, that may not make sense here).

binarystargames replied:
I had no idea about .IsNamedParams or scratch! I've refactored to do this as well as use both of these, those are really good options.

michaeltlombardi commented on layouts/partials/platen/docs/inject/head.html:
```diff
@@ -0,0 +1,2 @@
+{{- $platenStyles := resources.Get "platen.scss" | resources.ExecuteAsTemplate "platen.scss" . | resources.ToCSS | resources.Minify | resources.Fingerprint }}
+<link rel="stylesheet" href="{{ $platenStyles.RelPermalink }}" {{ template "integrity" $platenStyles }}>
```
Nit: missing trailing newline.
```suggestion
<link rel="stylesheet" href="{{ $platenStyles.RelPermalink }}" {{ template "integrity" $platenStyles }}>

```

michaeltlombardi commented on assets/platen.scss:
```diff
@@ -0,0 +1,7 @@
+    @media (min-width:56rem) {
+        flex-direction: row;
+    }
+}
```
Nit: missing trailing newline.
```suggestion
}

```

michaeltlombardi commented on layouts/partials/platen/docs/inject/head.html:
```diff
@@ -0,0 +1,2 @@
+{{- $platenStyles := resources.Get "platen.scss" | resources.ExecuteAsTemplate "platen.scss" . | resources.ToCSS | resources.Minify | resources.Fingerprint }}
+<link rel="stylesheet" href="{{ $platenStyles.RelPermalink }}" {{ template "integrity" $platenStyles }}>
```
We'll need to think about this a little bit - we don't need to _solve_ it here, but if we're going to inject our own stuff into the head, we need to provide theme users with a way to add their own injections without having to shadow ours and copy-paste our code in.

binarystargames replied:
That's true. I'm going to aim to refactor this to inject a new file via `baseof` because that one's already non-standard from hugo-book.

michaeltlombardi reviewed (changes requested):
Overall, this looks great! One minor note unrelated to the _code_, for this project if we can use rebase to update the working branch instead of merging upstream/main into it, that would be preferable - it helps keep the git history a little neater to follow.

It might also be possible to rebase and combine the first four commits into one - I think this work is really comprised of two chunks: Adding the new shortcode and extending the head injection.

Otherwise the implementation looks great and I'm stoked about this!

binarystargames commented:
Made a few changes as prescribed:

- Moved the inject out of head to baseof because we're already altering that from hugo-book.
- Refactored flexcolumns,html for improved readability/to use better tools as demonstrated

Definitely happy to rebase rather than merge in directly whenever this looks good.

michaeltlombardi commented on layouts/partials/platen/platen-head.html:
```diff
@@ -0,0 +1,2 @@
+{{- $platenStyles := resources.Get "platen.scss" | resources.ExecuteAsTemplate "platen.scss" . | resources.ToCSS | resources.Minify | resources.Fingerprint }}
+<link rel="stylesheet" href="{{ $platenStyles.RelPermalink }}" {{ template "integrity" $platenStyles }}>
```
Commenting here to remind myself:

I have come across this pattern a couple of times now, and I think it's probably worth defining a partial for ourselves so we can reference an SCSS style link injection via something like

```go
{{ partial "platen/GetScssLink" "platen.scss" }}
```

I've been using some `utils` partials over in toroidal - it turns out partials can be used to return values instead of writing text into a file, so that might be useful.

binarystargames replied:
Yeah, we used a similar value return with a partial with `platen/getCssClass` for class injection in `baseof`.

michaeltlombardi commented on layouts/shortcodes/flexcolumns.html:
```diff
@@ -0,0 +1,34 @@
+  </div>
+  {{- end -}}
+{{- end -}}
+</div>
```
A couple of (untested) suggestions:

```suggestion
{{-/*
  This shortcode can be used to turn markdown content into columns with a
  flexible ratio instead of being evenly distributed. It is a more powerful
  alternative to the `columns` shortcode.

  It has two optional parameters, `ratio` and `flattenInMobileView`. They can
  be specified by name or position. For example, these two shortcodes have the
  same functionality:

    {{< flexcolumns ratio="3:7" flattenInMobileView="false" >}}

    {{< flexcolumns "3:7" "false" >}}

  Note that to specify `flattenInMobileView` but not `ratio` without using
  named parameters, you **must** specify the ratio as an empty string:

    {{< flexcolumns "" "false" >}}
  
  The `ratio` parameter should be specified as a string of integers separated
  by colons (`:`), such as `1:2:1` or `1:1:3:1`. If `ratio` is not specified
  (or is set to an empty string, `""`), the columns are evenly distributed.
  For more information on flexible columns, see the docs on MDN:
  https://developer.mozilla.org/en-US/docs/Web/CSS/flex-grow
  
  If `flattenInMobileView` is specified as "true", the `flatten-in-mobile`
  class is added to the container for the columns. This causes the columns to
  collapse into a vertical stack when in mobile view to help make them more
  readable on smaller screens. If `flattenInMobileView` is not specified,
  the columns will still display horizontally, even on smaller screens.
*/-}}
{{- $ratio  := "" -}}
{{- $mobile := "" -}}

{{- if .IsNamedParams -}}
  {{- $ratio = (.Get "ratio") -}}
  {{- if eq (.Get "flattenInMobileView") "true" -}}
    {{- $mobile = " flatten-in-mobile" -}}
  {{- end -}}
{{- else -}}
  {{- $ratio = (.Get 0) -}}
  {{- if eq (.Get 1) "true" -}}
    {{- $mobile = " flatten-in-mobile" -}}
  {{- end -}}
{{- end -}}

{{- with $ratio -}}
  {{- $ratio = split . ":" -}}
{{- end -}}

<div class="book-columns flex flex-wrap{{ $mobile }}">
{{- if $ratio -}}
  {{- range $i, $val := split .Inner "<--->" -}}
    {{- if gt (len $ratio) $i}}
      <div style="flex-grow:{{ index $ratio $i }};" class="flex-even markdown-inner">
        {{- $val | $.Page.RenderString -}}
      </div>
    {{- end -}}
  {{- end -}}
{{- else -}}
  {{- range split .Inner "<--->" -}}
  <div class="flex-even markdown-inner">
    {{- . | $.Page.RenderString -}}
  </div>
  {{- end -}}
{{- end -}}
</div>
```

1. I don't _think_ I've done this so far, but I've started adding doc comments to the top of partials and shortcodes. My eventual goal is to add a little parser tool we can use to generate docs for our shortcodes/partials/etc from the first comment in a templated file. I'm also not _sure_ how the template will react with the code examples in the comment; it *should* be okay, but it might catch fire and needs manual testing.
2. While I did raise the point about `scratch` (I had been using global scratch, which is what you used here, but there's also the possibility of creating a local scratch with the [`newscratch` function][newscratch]), I think in this case there's actually some benefit to using local variables and setting them in the parameter-handling blocks:
   - It declares the parameters with their default values (in this case, empty strings) before anything else.
   - It allows us to immediately process the `flattenInMobileView` parameter to the correct value (`flatten-in-mobile` if set to `true`, empty string otherwise).
   - We can reference these multiply-used variables in the actual HTML generation without having to re-retrieve the value from the scratch map several times.
   - The use of the variables is more immediately clear to someone inspecting the template.
3. I slightly modified the class injection so that `$mobile` defaults to an empty string and is set to ` flatten-in-mobile` is set to keep from appending a trailing space to the class definition when the value is not used; this is not a big deal, but I was already poking at the code 😅 
4. I don't think the ratio splitting with the `with` directive is necessarily better than using `if` in this case, but I wanted to include it in case you weren't familiar with how `with` works; it's basically shorthand for "if this value is set to a nonzero value for this type, make it the context variable in the enclosed block" - so the inside becomes `$ratio = split . ":"` instead of `$ratio = split $ratio ":"`. Because `with` clobbers context, we can't use it in the div declaration without storing the page context in a variable; I also think changing it there would make reading the block _more_ confusing than helpful.



[newscratch]: https://gohugo.io/functions/scratch/#the-local-newscratch

binarystargames replied:
Tested this, and aside from one change (your big block comment didn't have a space on either side, so it threw an error! Load-bearing comments are my favorite kind of error) it seems to work. As such I've put it in because it's way clearer.

michaeltlombardi commented on layouts/shortcodes/flexcolumns.html:
```diff
@@ -0,0 +1,34 @@
+      <div style="flex-grow:{{ index $ratioSplit $i }};" class="flex-even markdown-inner">
+        {{- $val | $.Page.RenderString -}}
+      </div>
+    {{- end -}}
```
Can you help me understand this block?

If I understand correctly, the ratio is split into its pieces beforehand (i.e. the ratio `1:2:3` becomes the slice `[1, 2, 3]`) and `$i` is the current index of the markdown blocks split by `<--->`. So this conditional checks to see if `$i` is greater than the length of the slice of ratio values and, if so - only if so? - adds a `div` containing the rendered content with the `flex-grow` CSS attribute set to the content's index in the slice of ratio values.

What happens when the declared ratio is `1:2:3` and there's only two sections in the markdown? When there's four sections? My brain seems to think the first item in the split markdown (index `0`?) will always be less than the length of the slice of ratio values if they're non-zero - but your code is working, which means I'm definitely misunderstanding. 😭

binarystargames replied:
> So this conditional checks to see if $i is greater than the length of the slice of ratio values

gt works the other way: the slice length is greater than $i.

(Golang template scannability will be the death of me.)

michaeltlombardi replied:
Ahhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh

I _get_ why the functions are the way they are but I would kill for it to have logical operators implemented so it could be `if (len $ratio) > $i` 😭

michaeltlombardi reviewed (commented):
Some thoughts, suggestions, and questions. Also worth considering: if there are error conditions, can we handle those explicitly with the [`errorf`][errorf] function to give our users (who may not be technical at all) more helpful information than hugo's default when a template breaks?

[errorf]: https://gohugo.io/functions/errorf/

michaeltlombardi reviewed (approved):
Okay, this looks great! I think the only dangling (non-blocking) questions are:

- is if there's any sensible error handling we can use here - maybe to check the ratio string specified, make sure that it's numbers separated by colons?
- can the `flattenInMobileView` parameter be a direct boolean instead of a string?

binarystargames commented:
> sensible error handling

It's hard to believe, but this is actually pretty robust in that regard. The reason for this is how poorly formed CSS works: `flex-grow:(not a number/empty)` just doesn't get processed and falls back to `flex-grow:1`. 

At that same Alchemist link, you can see examples of:

- Needleburst: ratio="3::7"
- Immunity Shot: ratio="3:"
- Vitalizing Draught: ratio=":7
- Leadbody: ratio="3:a"
- Sliptongue: ratio="a:3"
- Incendiary Paste: ratio="3"
- Snake Venom: ratio="a"

The last two broke without a check for ":" so I've included a colon check. In all of the other cases, it just fell back to 1.

> can the `flattenInMobileView` parameter be a direct boolean instead of a string?

Sure can! The following now work:

- `{{< flexcolumns ratio="3:7" flattenInMobileView=true >}}`
- `{{< flexcolumns "3:7" true >}}`

michaeltlombardi commented on layouts/shortcodes/flexcolumns.html:
```diff
@@ -0,0 +1,65 @@
+      {{< flexcolumns "3:7" "false" >}}
+    Note that to specify `flattenInMobileView` but not `ratio` without using
+    named parameters, you **must** specify the ratio as an empty string:
+      {{< flexcolumns "" "false" >}}
```
```suggestion
      {{< flexcolumns ratio="3:7" flattenInMobileView=false >}}
      {{< flexcolumns "3:7" false >}}
    Note that to specify `flattenInMobileView` but not `ratio` without using
    named parameters, you **must** specify the ratio as an empty string:
      {{< flexcolumns "" false >}}
```

michaeltlombardi commented on layouts/shortcodes/flexcolumns.html:
```diff
@@ -0,0 +1,65 @@
+    For more information on flexible columns, see the docs on MDN:
+    https://developer.mozilla.org/en-US/docs/Web/CSS/flex-grow
+    
+    If `flattenInMobileView` is specified as "true", the `flatten-in-mobile`
```
```suggestion
    If `flattenInMobileView` is specified as `true`, the `flatten-in-mobile`
```

michaeltlombardi reviewed (approved):
Okay! That all makes sense and this looks *great*! Two minor fixes to the docs and my last (🤞🏻) question: With the way `flex-grow` works, I'm wondering if we should _warn_ when people submit an invalid ratio - probably just a single warning message like "Hey this ratio is invalid, we're going to fall back on an even ratio" and _maybe_ include an option to treat the warning as an error.

I couldn't quickly find the docs on turning warnings into errors, but I _did_ find how folks can [suppress an error](https://gohugo.io/functions/errorf/#suppress-errors). Not blocking for this PR, I'll merge tonight or tomorrow either way.
